netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -423,7 +824,450 @@ net_ebpf_extension_sock_addr_authorize_connection_classify(
+    UNREFERENCED_PARAMETER(flow_context);
+
+    filter_context = (net_ebpf_extension_sock_addr_wfp_filter_context_t*)filter->context;
+    ASSERT(filter_context != NULL);
+    if (filter_context == NULL)
+        goto Exit;
+
+    _net_ebpf_extension_sock_addr_copy_wfp_connection_fields(incoming_fixed_values, &sock_addr_ctx);
+
+    compartment_id = filter_context->compartment_id;
+    ASSERT((compartment_id == UNSPECIFIED_COMPARTMENT_ID) || (compartment_id == sock_addr_ctx.compartment_id));
+    if (compartment_id != UNSPECIFIED_COMPARTMENT_ID && compartment_id != sock_addr_ctx.compartment_id) {
+        // The client is not interested in this compartment Id.
+        NET_EBPF_EXT_LOG_MESSAGE_UINT32(
+            NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+            NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+            "The cgroup_sock_addr eBPF program is not interested in this compartmentId",
+            sock_addr_ctx.compartment_id);
+
+        verdict = BPF_SOCK_ADDR_VERDICT_PROCEED;
+        goto Exit;
+    }
+
+    // Get the connection context for this connection.
+    status = _net_ebpf_ext_get_connection_context(
+        incoming_metadata_values->transportEndpointHandle, &sock_addr_ctx, &connection_context);
+    if (!NT_SUCCESS(status)) {
+        // We did not find any connection context for this AUTH request. Block.
+        verdict = BPF_SOCK_ADDR_VERDICT_REJECT;
+        goto Exit;
+    }
+    verdict = connection_context->value.verdict;
+
+    _net_ebpf_ext_delete_connection_context(connection_context);
+    connection_context = NULL;
+
+Exit:
+    classify_output->actionType = (verdict == BPF_SOCK_ADDR_VERDICT_PROCEED) ? FWP_ACTION_PERMIT : FWP_ACTION_BLOCK;
+    // Clear FWPS_RIGHT_ACTION_WRITE only when it is a hard block.
+    if (classify_output->actionType == FWP_ACTION_BLOCK) {
+        classify_output->rights &= ~FWPS_RIGHT_ACTION_WRITE;
+    }
+
+    NET_EBPF_EXT_LOG_MESSAGE_UINT64_UINT64_UINT64(
+        NET_EBPF_EXT_TRACELOG_LEVEL_VERBOSE,
+        NET_EBPF_EXT_TRACELOG_KEYWORD_SOCK_ADDR,
+        "authorize_connection_classify",
+        incoming_metadata_values->transportEndpointHandle,
+        sock_addr_ctx.protocol,
+        verdict);
+
+    return;
+}
+
+/*
+ * Default action is BLOCK. If this callout is being invoked, it means at least one
+ * eBPF program is attached. Hence no connection should be allowed unless allowed by
+ * the eBPF program.
+ *
+ * If the eBPF program verdict is BLOCK, no context is created and AUTH classify will
+ * block the connection by default.
+ */
+void
+net_ebpf_extension_sock_addr_redirect_connection_classify(
+    _In_ const FWPS_INCOMING_VALUES* incoming_fixed_values,
+    _In_ const FWPS_INCOMING_METADATA_VALUES* incoming_metadata_values,
+    _Inout_opt_ void* layer_data,
+    _In_opt_ const void* classify_context,
+    _In_ const FWPS_FILTER* filter,
+    uint64_t flow_context,
+    _Inout_ FWPS_CLASSIFY_OUT* classify_output)
+{
+    uint32_t verdict;
```

Comment:
Function too long. Consider breaking into smaller functions no longer than 2 screen sizes.

Reply:
Have somewhat refactored the code.